pkg/api/validate/dynamic/dynamic.go

```diff
@@ -781,6 +783,61 @@ func (dv *dynamic) ValidateSubnets(ctx context.Context, oc *api.OpenShiftCluster
 	return nil
 }
 
+func (dv *dynamic) ValidatePreConfiguredNSGs(ctx context.Context, oc *api.OpenShiftCluster, subnets []Subnet) error {
+	dv.log.Print("ValidatePreConfiguredNSGs")
+
+	if oc.Properties.NetworkProfile.PreconfiguredNSG != api.PreconfiguredNSGEnabled {
+		return nil // exit early
+	}
+
+	subnetByID, err := dv.createSubnetMapByID(ctx, subnets)
+	if err != nil {
+		return err
+	}
+
+	for _, s := range subnetByID {
+		nsgID := s.NetworkSecurityGroup.ID
+		if nsgID == nil || *nsgID == "" {
+			return api.NewCloudError(
+				http.StatusBadRequest,
+				api.CloudErrorCodeNotFound,
+				"",
+				errMsgNSGNotProperlyAttached,
+			)
+		}
+
+		nsg, err := azure.ParseResourceID(*nsgID)
+		if err != nil {
+			return err
+		}
+
+		err = dv.validateActions(ctx, &nsg, []string{
+			"Microsoft.Network/networkSecurityGroups/join/action",
+		})
+
+		if err == wait.ErrWaitTimeout {
+			errCode := api.CloudErrorCodeInvalidResourceProviderPermissions
+			if dv.authorizerType == AuthorizerClusterServicePrincipal {
+				errCode = api.CloudErrorCodeInvalidServicePrincipalPermissions
+			}
+			return api.NewCloudError(
+				http.StatusBadRequest,
+				errCode,
+				"",
+				errMsgSPHasNoRequiredPermissionsOnNSG,
+				dv.authorizerType,
+				dv.appID,
+				*nsgID,
+			)
+		}
+		if err != nil {
+			return err
+		}
```

Comment:
can you put this into a function please?

Reply:
@facchettos done